balancer/base/hierarchy/hierarchy.go

```diff
@@ -0,0 +1,98 @@
+/*
+ *
+ * Copyright 2020 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package hierarchy contains functions to set and get hierarchy string from
+// addresses.
+package hierarchy
+
+import (
+	"google.golang.org/grpc/attributes"
+	"google.golang.org/grpc/resolver"
+)
+
+type hierarchicalPathKeyType string
+
+const hierarchicalPathKey = hierarchicalPathKeyType("grpc.internal.address.hierarchical_path")
+
+// Get returns the hierarchical path of addr.
+func Get(addr resolver.Address) []string {
+	attrs := addr.Attributes
+	if attrs == nil {
+		return nil
+	}
+	path, ok := attrs.Value(hierarchicalPathKey).([]string)
+	if !ok {
+		return nil
+	}
+	return path
+}
+
+// Set overrides the hierarchical path in addr with path.
+func Set(addr resolver.Address, path []string) resolver.Address {
```

Comment:
What do you think about accepting a pointer and modifing in place?

Reply:
Copying is cleaner to use?

```go
// Copy
newAddrs = append(newAddrs, Set(addr, ["a","b"]))

// Update
Set(addr, ["a","b"])
newAddrs = append(newAddrs, addr)

// With update, or even do everything in place?
for _, addr := range addrs {
    Set(addr, ["a","b"])
}
```

Another question is, what should `Group()` do? Should it modify the passed in addresses, or copy?